File: src/domain/battleConfigMacros.ts
export type BattleNumberFieldKey =
  | "maxRounds"
  | "minimumDamage"
  | "randomSeed"
  | "armorFormulaBase"
  | "maxArmorDamageReduction"
  | "elementAdvantageDamageRate"
  | "elementDisadvantageDamageRate";

interface BattleConfigNumberMacroDefinition {
  key: BattleNumberFieldKey;
  label: string;
  default: number;
  min: number;
  max?: number;
  step: number;
  suffix?: string;
}

interface BattleConfigMacroDocument {
  requiredRule: string;
  battleNumberFields: BattleConfigNumberMacroDefinition[];
}

async function loadBattleConfigMacroDocument() {
  const configUrl = new URL("../config/battle-config-macros.json", import.meta.url);

  if (typeof window === "undefined") {
    const { readFile } = await import("node:fs/promises");
    const content = await readFile(configUrl, "utf8");
    return JSON.parse(content) as BattleConfigMacroDocument;
  }

  const response = await fetch(configUrl);
  if (!response.ok) {
    throw new Error(`加载整场战斗参数定义失败：HTTP ${response.status}`);
  }

  return (await response.json()) as BattleConfigMacroDocument;
}

function validateBattleConfigMacroDocument(document: BattleConfigMacroDocument) {
  const keySet = new Set<string>();

  for (const macro of document.battleNumberFields) {
    if (keySet.has(macro.key)) {
      throw new Error(`整场战斗参数定义存在重复 key: ${macro.key}`);
    }

    keySet.add(macro.key);
  }

  return document;
}

const battleConfigMacroDocument = validateBattleConfigMacroDocument(await loadBattleConfigMacroDocument());

export const battleConfigMacroRule = battleConfigMacroDocument.requiredRule;
export const battleConfigNumberMacros = Object.freeze(battleConfigMacroDocument.battleNumberFields);
export const battleConfigNumberMacroMap = Object.freeze(
  Object.fromEntries(battleConfigNumberMacros.map((macro) => [macro.key, macro])) as Record<
    BattleNumberFieldKey,
    BattleConfigNumberMacroDefinition
  >,
);
export const battleNumberDefaults = Object.freeze(
  Object.fromEntries(battleConfigNumberMacros.map((macro) => [macro.key, macro.default])) as Record<
    BattleNumberFieldKey,
    number
  >,
);

export function formatBattleConfigValue(field: BattleNumberFieldKey, value: number) {
  const macro = battleConfigNumberMacroMap[field];
  return `${value}${macro.suffix ?? ""}`;
}

File: backend/battle_config_macros.py
from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import NotRequired, TypedDict, cast


ROOT_DIR = Path(__file__).resolve().parent.parent
BATTLE_CONFIG_MACRO_PATH = ROOT_DIR / "src" / "config" / "battle-config-macros.json"


class BattleConfigNumberMacroDefinition(TypedDict):
    key: str
    label: str
    default: int | float
    defaultFactory: NotRequired[str]
    min: int | float
    max: NotRequired[int | float]
    step: int | float
    suffix: NotRequired[str]


class BattleConfigMacroDocument(TypedDict):
    requiredRule: str
    battleNumberFields: list[BattleConfigNumberMacroDefinition]


@lru_cache(maxsize=1)
def load_battle_config_macro_document() -> BattleConfigMacroDocument:
    document = cast(BattleConfigMacroDocument, json.loads(BATTLE_CONFIG_MACRO_PATH.read_text(encoding="utf-8")))
    key_set: set[str] = set()

    for macro in document["battleNumberFields"]:
        key = str(macro["key"])
        if key in key_set:
            raise ValueError(f"整场战斗参数定义存在重复 key: {key}")

        key_set.add(key)

    return document


def get_battle_config_macro_rule() -> str:
    return str(load_battle_config_macro_document()["requiredRule"])


def get_battle_number_macros() -> list[BattleConfigNumberMacroDefinition]:
    return list(load_battle_config_macro_document()["battleNumberFields"])


def get_battle_number_defaults() -> dict[str, int | float]:
    return {str(macro["key"]): macro["default"] for macro in load_battle_config_macro_document()["battleNumberFields"]}
